docs/design/operator.md

```diff
@@ -0,0 +1,387 @@
+```
+
+### CephCSIDriver CRD
+
+Manages the installation, lifecycle management, and configuration for CephFS,
+RBD, and NFS CSI drivers within namespaces.
+
+- Allows customization of driver settings on a per-driver basis.
+- Best practice dictates only a single instance of RBD/NFS/CephFS drivers
+  should be created in a single namespace. If there is a need to deploy a
+  second instance of a driver of a similar type, it should be deployed in a
+  different namespace.
+- Each CSI driver must have a unique name across all namespaces.
+
+```yaml
+---
+kind: CephCSIDriver
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: "<prefix>.<driver_type>.csi.ceph.com"
+  namespace: <operator-namespace>
+spec:
+  fsGroupPolicy: File
+  encryption:
+    configMapRef:
+      name: encryption-config-map-name
+  plugin:
+    priorityClassName: system-node-critical
+    updateStrategy:
+      type: RollingUpdate
+      rollingUpdate:
+        maxUnavailable: 1
+    labels:
+      app: cephfs-plugin
+    annotations:
+      k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+  provisioner:
+    labels:
+      app: ceph-fs-provisioner
+    annotations:
+      k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+    provisionerReplicas: 2
+    leaderElection:
+      leaseDuration: 100
+      renewDeadline: 100
+      retryPeriod: 10
+    attachRequired: true
+  liveness:
+    metricsPort: 8000
+  deployCSIAddons: false
+  kernelClient: true
+  kernelMountOptions: ms_mode=secure
+status:
+  phase: Failed
+  reason: csi driver with same name already exists in the cluster
+```
+
+### CephCSICephConfig CRD
+
+Stores connection and configuration details for a single Ceph cluster and
+provide the information to be used by multiple CSI drivers.
+
+```yaml
+---
+kind: CephCSICephCluster
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: ceph-cluster-1
+  namespace: <operator-namespace>
+spec:
+  monitors:
+  - 10.98.44.171:6789
+  - 10.98.44.172:6789
+  - 10.98.44.173:6789
+  ReadAffinity:
+    crushLocationLabels:
+    - kubernetes.io/hostname
+    - topology.kubernetes.io/region
+    - topology.kubernetes.io/zone
+  rbdMirrorDaemonCount: 2
+  cephConfig:
+    global:
+      auth_cluster_required: none
+      auth_service_required: none
+      auth_client_required: none
+      rbd_validate_pool: false
+status: {}
+```
+
+### CephCSIConfig CRD
+
+Contains details about CephFS, RBD, and NFS configuration to be used when
+communicating with Ceph. Include a reference to a CephCSICephCluster holding
+the connection information for the target Ceph cluster.
+
+```yaml
+---
+kind: CephCSIConfig
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: storage
+  namespace: <operator-namespace>
+spec:
+  cephClusterRef:
+    name: ceph-cluster-1
+  cephFS:
+    subvolumeGroup: csi
+    kernelMountOptions: readdir_max_bytes=1048576,norbytes
+    fuseMountOptions: debug
+  rbd:
+    radosNamespace: rados-test
+status:
+  phase: Succeeded
+  reason: successfully linked to CephClusterRef
+```
```

Comment:
Why is this not part of CephCSIDriver CRD?

Reply:
thats because a single cephcsi driver name have multiple ceph clusters created and its not required when we are starting a cephcsi driver, these details can be added later any point of time.